PyInstaller/lib/modulegraph/modulegraph.py

```diff
@@ -997,6 +996,36 @@ def visit_Expression(self, node):
     visit_Call = visit_Expression
 
 
+class Engine(object):
+    """
+    This object removes the complexity needed for iterative behavior
+    out of modulegraph.
+    """
+
+    def __init__(self, processor):
+        assert callable(processor)
+
+        self.actions = deque()
+        self.register = deque()
+        self.processor = processor
+
+    def put(self, action):
+        if sys.version_info < (3, 0):
+            self.processor(*action)
```

Comment:
Why is Py < 3.0 handled differently? What are the implications=

Reply:
Python 2.7 is handled differently because `deque` performance is so bad that it led to a test timeout. The simplest solution to the problem is to do what we previously did, which is use recursion on Python 2.7.